coala_langserver/langserver.py

```diff
@@ -3,84 +3,76 @@
 import socketserver
 import traceback
 
-from .jsonrpc import JSONRPC2Connection, ReadWriter, TCPReadWriter
+from pyls.jsonrpc.endpoint import Endpoint
+from pyls.jsonrpc.dispatchers import MethodDispatcher
+from pyls.jsonrpc.streams import JsonRpcStreamReader
+from pyls.jsonrpc.streams import JsonRpcStreamWriter
+from coala_utils.decorators import enforce_signature
 from .log import log
 from .coalashim import run_coala_with_specific_file
 from .uri import path_from_uri
 from .diagnostic import output_to_diagnostics
 
 
-class ThreadingTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
-    pass
+class _StreamHandlerWrapper(socketserver.StreamRequestHandler, object):
+    """
+    A wrapper class that is used to construct a custom handler class.
+    """
 
+    delegate = None
 
-class LangserverTCPTransport(socketserver.StreamRequestHandler):
+    def setup(self):
+        super(_StreamHandlerWrapper, self).setup()
+        self.delegate = self.DELEGATE_CLASS(self.rfile, self.wfile)
 
     def handle(self):
-        s = LangServer(conn=TCPReadWriter(self.rfile, self.wfile))
-        try:
-            s.listen()
-        except Exception as e:
-            tb = traceback.format_exc()
-            log('ERROR: {} {}'.format(e, tb))
+        self.delegate.start()
 
 
-class LangServer(JSONRPC2Connection):
+class LangServer(MethodDispatcher):
     """
     Language server for coala base on JSON RPC.
     """
 
-    def __init__(self, conn=None):
-        super().__init__(conn=conn)
+    def __init__(self, rx, tx):
         self.root_path = None
+        self._jsonrpc_stream_reader = JsonRpcStreamReader(rx)
+        self._jsonrpc_stream_writer = JsonRpcStreamWriter(tx)
+        self._endpoint = Endpoint(self, self._jsonrpc_stream_writer.write)
+        self._dispatchers = []
+        self._shutdown = False
 
-    def handle(self, _id, request):
-        """
-        Handle the request from language client.
-        """
-        log('REQUEST: ', request)
-        resp = None
-
-        if request['method'] == 'initialize':
-            resp = self.serve_initialize(request)
-        # TODO: Support did_change and did_change_watched_files.
-        # elif request["method"] == "textDocument/didChange":
-        #     resp = self.serve_change(request)
-        # elif request["method"] == "workspace/didChangeWatchedFiles":
-        #     resp = self.serve_did_change_watched_files(request)
-        elif request['method'] == 'textDocument/didSave':
-            self.serve_did_save(request)
-
-        if resp is not None:
-            self.write_response(request['id'], resp)
-
-    def serve_initialize(self, request):
+    def start(self):
+        self._jsonrpc_stream_reader.listen(self._endpoint.consume)
+
+    def m_initialize(self, rootUri=None, rootPath=None, **kargs):
```

Comment:
`def m_initialize(self, **params)`

then nothing below needs to be modified, except removal of `params = request['params']`

sorry I didnt notice this earlier.

this helps show the similarities of the two jsonrpc's , making differences more clear